File: mobile/android/app/src/main/kotlin/com/masim05/bloodpressure/mobile/core/validation/Validators.kt
package com.masim05.bloodpressure.mobile.core.validation

import com.masim05.bloodpressure.mobile.core.model.HistoryFilter
import com.masim05.bloodpressure.mobile.core.model.MeasurementImage

sealed class ValidationResult {
    data object Valid : ValidationResult()
    data class Invalid(val reason: ValidationError) : ValidationResult()
}

enum class ValidationError {
    InvalidEmail,
    InvalidPassword,
    InvalidImage,
    InvalidDate,
    DateOrder,
}

object Validators {
    private val emailPattern = Regex("^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$")
    private val datePattern = Regex("^\\d{4}-\\d{2}-\\d{2}$")

    fun email(value: String): ValidationResult =
        if (emailPattern.matches(value.trim())) ValidationResult.Valid else ValidationResult.Invalid(ValidationError.InvalidEmail)

    fun password(value: String): ValidationResult =
        if (value.length >= 8) ValidationResult.Valid else ValidationResult.Invalid(ValidationError.InvalidPassword)

    fun image(image: MeasurementImage): ValidationResult =
        if (image.sizeBytes in 1..10_485_760 && image.mimeType in setOf("image/jpeg", "image/png")) {
            ValidationResult.Valid
        } else {
            ValidationResult.Invalid(ValidationError.InvalidImage)
        }

    fun historyFilter(filter: HistoryFilter): ValidationResult {
        if (filter.from.isNotBlank() && !datePattern.matches(filter.from)) {
            return ValidationResult.Invalid(ValidationError.InvalidDate)
        }
        if (filter.to.isNotBlank() && !datePattern.matches(filter.to)) {
            return ValidationResult.Invalid(ValidationError.InvalidDate)
        }
        if (filter.from.isNotBlank() && filter.to.isNotBlank() && filter.from > filter.to) {
            return ValidationResult.Invalid(ValidationError.DateOrder)
        }
        return ValidationResult.Valid
    }
}

File: mobile/ios/Blood pressure tracker/Blood pressure tracker/Validators.swift
// Validators.swift
// Blood pressure tracker
// Input validation mirroring Android's Validators.kt

import Foundation

// MARK: - Validation types

enum ValidationResult: Equatable {
    case valid
    case invalid(ValidationError)
}

enum ValidationError: Equatable {
    case invalidEmail
    case invalidPassword
    case invalidImage
    case invalidDate
    case dateOrder
}

// MARK: - Validators

enum Validators {
    private static let emailPattern: NSRegularExpression = {
        try! NSRegularExpression(pattern: #"^[^@\s]+@[^@\s]+\.[^@\s]+$"#)
    }()

    private static let datePattern: NSRegularExpression = {
        try! NSRegularExpression(pattern: #"^\d{4}-\d{2}-\d{2}$"#)
    }()

    static func email(_ value: String) -> ValidationResult {
        let trimmed = value.trimmingCharacters(in: .whitespaces)
        let range = NSRange(trimmed.startIndex..., in: trimmed)
        return emailPattern.firstMatch(in: trimmed, range: range) != nil
            ? .valid
            : .invalid(.invalidEmail)
    }

    static func password(_ value: String) -> ValidationResult {
        value.count >= 8 ? .valid : .invalid(.invalidPassword)
    }

    static func image(_ image: MeasurementImage) -> ValidationResult {
        let validMimeTypes: Set<String> = ["image/jpeg", "image/png"]
        if image.sizeBytes >= 1 && image.sizeBytes <= 10_485_760 && validMimeTypes.contains(image.mimeType) {
            return .valid
        }
        return .invalid(.invalidImage)
    }

    static func historyFilter(_ filter: HistoryFilter) -> ValidationResult {
        if !filter.from.isEmpty {
            let range = NSRange(filter.from.startIndex..., in: filter.from)
            if datePattern.firstMatch(in: filter.from, range: range) == nil {
                return .invalid(.invalidDate)
            }
        }
        if !filter.to.isEmpty {
            let range = NSRange(filter.to.startIndex..., in: filter.to)
            if datePattern.firstMatch(in: filter.to, range: range) == nil {
                return .invalid(.invalidDate)
            }
        }
        if !filter.from.isEmpty && !filter.to.isEmpty && filter.from > filter.to {
            return .invalid(.dateOrder)
        }
        return .valid
    }
}
